cloudwatch-integ/kvsWebRTCClientMasterCloudwatch.cpp

```diff
@@ -0,0 +1,521 @@
+                    CHK_STATUS(writeFirstFrameSentTimeToFile(tsFileName));
+                    PROFILE_WITH_START_TIME_OBJ(pSampleConfiguration->sampleStreamingSessionList[i]->offerReceiveTime, firstFrameTime, "Time to first frame");
+                    CppInteg::Cloudwatch::getInstance().monitoring.pushTimeToFirstFrame(firstFrameTime,
+                                                                                        Aws::CloudWatch::Model::StandardUnit::Milliseconds);
+
+                    calculateDisconnectToFrameSentTime(pSampleConfiguration);
+                    pSampleConfiguration->sampleStreamingSessionList[i]->firstFrame = FALSE;
+                }
+            } else {
+                // Reset file index to stay in sync with video frames.
+                fileIndex = 0;
+            }
+        }
+        MUTEX_UNLOCK(pSampleConfiguration->streamingSessionListReadLock);
+        THREAD_SLEEP(SAMPLE_AUDIO_FRAME_DURATION);
+    }
+
+    CleanUp:
+    DLOGI("[KVS Master] closing audio thread");
+    return (PVOID) (ULONG_PTR) retStatus;
+}
+
+INT32 main(INT32 argc, CHAR* argv[])
+{
+    STATUS retStatus = STATUS_SUCCESS;
+    UINT32 frameSize;
+    PSampleConfiguration pSampleConfiguration = NULL;
+    PCHAR region;
+    UINT32 terminateTimerId = MAX_UINT32;
+    CHAR channelName[MAX_CHANNEL_NAME_LEN];
+    PCHAR channelNamePrefix;
+    Aws::SDKOptions options;
+    options.httpOptions.installSigPipeHandler = TRUE;
+    CHAR tsFileName[MAX_PATH_LEN + 1];
+    TID profilingThread;
+    Aws::InitAPI(options);
+    {
+        SET_INSTRUMENTED_ALLOCATORS();
+        UINT32 logLevel = setLogLevel();
+        // Initialize KVS WebRTC. This must be done before anything else, and must only be done once.
+        CHK_STATUS(initKvsWebRtc());
+
+        if (USE_IOT) {
+            PCHAR pChannelName;
+            CHK_ERR((pChannelName = GETENV(IOT_CORE_THING_NAME)) != NULL, STATUS_INVALID_OPERATION, "AWS_IOT_CORE_THING_NAME must be set since USE_IOT is enabled");
+            STRNCPY(channelName, pChannelName, SIZEOF(channelName));
+        } else {
+            channelNamePrefix = argc > 1 ? argv[1] : CHANNEL_NAME_PREFIX;
+            SNPRINTF(channelName, SIZEOF(channelName), CHANNEL_NAME_TEMPLATE, channelNamePrefix, RUNNER_LABEL);
+        }
+        CHK_STATUS(createSampleConfiguration(channelName, SIGNALING_CHANNEL_ROLE_TYPE_MASTER, TRUE, TRUE, logLevel, &pSampleConfiguration));
+        CHK_STATUS(setUpCredentialProvider(pSampleConfiguration, USE_IOT));
+
+        pSampleConfiguration->forceTurn = FORCE_TURN_ONLY;
+        pSampleConfiguration->enableMetrics = ENABLE_METRICS;
+        pSampleConfiguration->channelInfo.useMediaStorage = ENABLE_STORAGE;
+        pSampleConfiguration->audioCodec = AUDIO_CODEC;
+        pSampleConfiguration->videoCodec = VIDEO_CODEC;
+
+        if(pSampleConfiguration->channelInfo.useMediaStorage) {
+            pSampleConfiguration->audioSource = sendRealAudioPackets;
```

Comment:
What's the reason for using real source only when storage is enabled?

Reply:
Media server backend does not allow dummy frames.